src/serde/ser.rs

```diff
@@ -204,172 +210,206 @@ where
-        value.serialize(&mut *self)
+        value.serialize(ValueSerializer::new(
+            self.value_writer.with_annotations([name, variant])?,
+        ))
     }
 
     fn serialize_seq(self, _len: Option<usize>) -> Result<Self::SerializeSeq, Self::Error> {
-        self.writer.step_in(IonType::List)?;
-        Ok(self)
+        Ok(SeqWriter {
+            seq_writer: self.value_writer.list_writer()?,
+        })
     }
 
     fn serialize_tuple(self, _len: usize) -> Result<Self::SerializeTuple, Self::Error> {
-        self.writer.step_in(IonType::List)?;
-        Ok(self)
+        Ok(SeqWriter {
+            seq_writer: self.value_writer.list_writer()?,
+        })
     }
 
     fn serialize_tuple_struct(
         self,
-        _name: &'static str,
+        name: &'static str,
         _len: usize,
     ) -> Result<Self::SerializeTupleStruct, Self::Error> {
-        self.writer.step_in(IonType::List)?;
-        Ok(self)
+        Ok(SeqWriter {
+            seq_writer: self.value_writer.with_annotations(name)?.list_writer()?,
+        })
     }
 
     fn serialize_tuple_variant(
         self,
-        _name: &'static str,
+        name: &'static str,
         _variant_index: u32,
         variant: &'static str,
         _len: usize,
     ) -> Result<Self::SerializeTupleVariant, Self::Error> {
-        self.writer.set_annotations(vec![variant]);
-        self.writer.step_in(IonType::List)?;
-        Ok(self)
+        Ok(SeqWriter {
+            seq_writer: self
+                .value_writer
+                .with_annotations([name, variant])?
+                .list_writer()?,
+        })
     }
 
     fn serialize_map(self, _len: Option<usize>) -> Result<Self::SerializeMap, Self::Error> {
-        self.writer.step_in(IonType::Struct)?;
-        Ok(self)
+        Ok(MapWriter {
+            map_writer: self.value_writer.struct_writer()?,
+        })
     }
 
     fn serialize_struct(
         self,
         _name: &'static str,
         _len: usize,
     ) -> Result<Self::SerializeStruct, Self::Error> {
-        self.writer.step_in(IonType::Struct)?;
-        Ok(self)
+        Ok(MapWriter {
+            map_writer: self.value_writer.struct_writer()?,
+        })
     }
 
     fn serialize_struct_variant(
         self,
-        _name: &'static str,
+        name: &'static str,
         _variant_index: u32,
         variant: &'static str,
         _len: usize,
     ) -> Result<Self::SerializeStructVariant, Self::Error> {
-        self.writer.set_annotations(vec![variant]);
-        self.writer.step_in(IonType::Struct)?;
-        Ok(self)
+        Ok(MapWriter {
+            map_writer: self
+                .value_writer
+                .with_annotations([name, variant])?
+                .struct_writer()?,
+        })
     }
 }
 
-impl<'a, E> ser::SerializeSeq for &'a mut Serializer<E>
-where
-    E: IonWriter,
-{
+pub struct SeqWriter<V: ValueWriter> {
+    seq_writer: V::ListWriter,
+}
```

Comment:
🗺️ This type and `MapWriter` below exist because it's not possible to implement a trait (in this case: `ser::SerializeSeq` and `ser::SerializeMap`) for a generic like `V::ListWriter` because it violates [the orphan rules](https://doc.rust-lang.org/reference/items/implementations.html#orphan-rules).

These shim types are locally defined, which means that we're not also implementing the trait for other `V: ValueWriter` types we don't own. (Though `ValueWriter` is effectively sealed, sealed traits aren't an official feature and so can't be considered when resolving the orphan rules.)